File: rb/src/day02.rb
require_relative 'util'

def part1(input)
  horiz, depth = 0,0
  input.lines do |line|
    dir, num = line.split(/ /)
    num = num.to_i
    case dir
    when 'up'
      depth -= num
    when 'down'
      depth += num
    when 'forward'
      horiz += num
    else
      raise '?'
    end
  end
  ans = horiz * depth
  puts ans
end

def part2(input)
  horiz, depth, aim = 0,0,0
  input.lines do |line|
    dir, num = line.split(/ /)
    num = num.to_i
    case dir
    when 'up'
      aim -= num
    when 'down'
      aim += num
    when 'forward'
      horiz += num
      depth += aim * num
    else
      raise '?'
    end
  end
  ans = horiz * depth
  puts ans
end

with_input do |input|
  part1(input)
  part2(input)
end

File: ts/src/day02.ts
import * as util from "./util"

function part1(input: string) {
  let ans: number = 0
  let horiz = 0
  let depth = 0
  input.split(/\n/).map(line => {
    let _, [dir, num] = line.split(/ /)
    let n = Number(num)

    if (dir === "up") {
      depth -= n
    } else if (dir === "down") {
      depth += n
    } else if (dir === "forward") {
      horiz += n
    } else {
      throw `Invalid ${line}`
    }
  })

  ans = horiz * depth
  console.log(ans)
}

function part2(input: string) {
  let ans: number = 0
  let horiz = 0
  let depth = 0
  let aim = 0
  input.split(/\n/).map(line => {
    let _, [dir, num] = line.split(/ /)
    let n = Number(num)

    if (dir === "up") {
      aim -= n
    } else if (dir === "down") {
      aim += n
    } else if (dir === "forward") {
      horiz += n
      depth += aim * n
    } else {
      throw `Invalid ${line}`
    }
  })

  ans = horiz * depth
  console.log(ans)
}

util.withInput("input/day02.txt", input => {
  part1(input)
  part2(input)
})